feat: new sequelize dialect for DuckDB

elefeint opened the pull request:
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions? No new tests, but all existing tests for the new dialect pass.
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? Not yet but I will; it seems to be useful to link back to this PR once it's open.
- [x] Did you update the typescript typings accordingly (if applicable)? No new public types were introduced.
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving? No
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)? yes


## Description of Changes

This PR contributes a new DuckDB dialect, compatible with both DuckDB and MotherDuck. DuckDB is an in-process OLAP database. MotherDuck is a cloud data warehouse based on and compatible with DuckDB. The dialect is based on `duckdb-async` DuckDB Node API.

I am happy to break this PR up into smaller ones, if that's helpful.

All unit and integration tests pass.

*Limitations:*
* indexes, unique constraints and foreign keys are not supported in this dialect. DuckDB technically supports both; however the presence of indexes means the data under index cannot be updated, which would cause unexpected issues for end users. 
* for similar reasons, the dialect only supports the RETURNING clause on INSERT queries, but not on UPDATE queries.
* DuckDB does not support adding or dropping constraints on an existing table
* DuckDB does not support transaction savepoints or transaction isolation levels


*Changes in this PR outside of the `packages/duckdb` dialect folder:*
* The changes in `core/src` are limited to where there already was an exclusion for another dialect.
* the changes in `core/test` are limited to
	* gating tests that rely on `constraints.foreignKey`, `constraints.unique`, `isolationLevels`
	* making implied `findAll` order explicit since the test relies on the ordering (`integration/associations/multiple-level-filters.test.js`).
	* accounting for duckdb CHAR type being just an alias of VARCHAR with none of the CHAR[length] limitations (data-types.test.ts)
	
## List of Breaking Changes
n/a

WikiRik commented:
Thanks for working on this! Unfortunately due to the limited resources the maintainers have available and priorities that we have towards landing Oracle support in v7-alphas, making the new CLI feature complete and in general finishing v7. That means that it will probably be a while before we review this PR properly. Just to convey the expectations quite early so you don't get disappointed

elefeint commented:
That's no problem -- we have a customer who requested the dialect, and they can test it in the meantime from the forked repo. I'll clean up the linter and send a documentation PR, so it's ready for review when you get to it.

elefeint commented:
I've added a documentation PR at https://github.com/sequelize/website/pull/775 and fixed the linter issues. 
`yarn test:format` passes locally now, but on CI there seems to be an infrastructural issue; I am not sure if you've seen that before:
```
Error: [prettier] [error] Unable to read file "install-build-node-20.tar":
Error: [prettier] [error] Invalid string length
```

dazhuangc commented:
> resources the maintainers have available and priorities that we have towards landing Oracle support in v7-alphas, making the new CLI feature complete and in general finishing v7. That means that it will probably be a while before we review this PR properly. Just to convey the ex

I am facing the same issue in https://github.com/sequelize/sequelize/pull/17758

dazhuangc commented:
> I've added a documentation PR at [sequelize/website#775](https://github.com/sequelize/website/pull/775) and fixed the linter issues. `yarn test:format` passes locally now, but on CI there seems to be an infrastructural issue; I am not sure if you've seen that before:
> 
> ```
> Error: [prettier] [error] Unable to read file "install-build-node-20.tar":
> Error: [prettier] [error] Invalid string length
> ```

I am facing the same issue in https://github.com/sequelize/sequelize/pull/17758

elefeint commented:
@WikiRik What do you think the timeline looks like for the DuckDB dialect? (I fully understand that this is a huge PR -- would it help to break it into pieces? For example, the changes to `core/src` are minimal).

elefeint commented:
Closing in favor of #18080
